File: Utils/src/main/java/org/tobi29/scapes/engine/utils/io/ReadableByteStream.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.utils.io;

import org.tobi29.scapes.engine.utils.BufferCreator;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

public interface ReadableByteStream {
    int remaining();

    default boolean hasRemaining() {
        return remaining() > 0;
    }

    default void skip(int len) throws IOException {
        get(BufferCreator.bytes(len));
    }

    default ReadableByteStream get(ByteBuffer buffer) throws IOException {
        return get(buffer, buffer.remaining());
    }

    ReadableByteStream get(ByteBuffer buffer, int len) throws IOException;

    default boolean getSome(ByteBuffer buffer) throws IOException {
        return getSome(buffer, buffer.remaining());
    }

    boolean getSome(ByteBuffer buffer, int len) throws IOException;

    default ReadableByteStream get(byte[] src) throws IOException {
        return get(src, 0, src.length);
    }

    default ReadableByteStream get(byte[] src, int off, int len)
            throws IOException {
        return get(ByteBuffer.wrap(src, off, len));
    }

    default int getSome(byte[] src, int off, int len) throws IOException {
        ByteBuffer buffer = ByteBuffer.wrap(src, off, len);
        boolean available = getSome(buffer);
        int position = buffer.position();
        if (position == 0 && !available) {
            return -1;
        }
        return position;
    }

    default boolean getBoolean() throws IOException {
        return get() != 0;
    }

    byte get() throws IOException;

    default short getUByte() throws IOException {
        short value = get();
        if (value < 0) {
            value += 0x100;
        }
        return value;
    }

    short getShort() throws IOException;

    default int getUShort() throws IOException {
        int value = getShort();
        if (value < 0) {
            value += 0x10000;
        }
        return value;
    }

    int getInt() throws IOException;

    default long getUInt() throws IOException {
        long value = getInt();
        if (value < 0) {
            value += 0x100000000L;
        }
        return value;
    }

    long getLong() throws IOException;

    float getFloat() throws IOException;

    double getDouble() throws IOException;

    default byte[] getByteArray() throws IOException {
        return getByteArray(Integer.MAX_VALUE);
    }

    default byte[] getByteArray(int limit) throws IOException {
        int len = getUByte();
        if (len == 0xFF) {
            len = getInt();
        }
        if (len < 0 || len > limit) {
            throw new IOException(
                    "Array length outside of 0 to " + limit + ": " + limit);
        }
        byte[] array = new byte[len];
        get(array);
        return array;
    }

    default byte[] getByteArrayLong() throws IOException {
        return getByteArrayLong(Integer.MAX_VALUE);
    }

    default byte[] getByteArrayLong(int limit) throws IOException {
        int len = getUShort();
        if (len == 0xFFFF) {
            len = getInt();
        }
        if (len < 0 || len > limit) {
            throw new IOException(
                    "Array length outside of 0 to " + limit + ": " + limit);
        }
        byte[] array = new byte[len];
        get(array);
        return array;
    }

    default String getString() throws IOException {
        return getString(Integer.MAX_VALUE);
    }

    default String getString(int limit) throws IOException {
        return new String(getByteArray(limit), StandardCharsets.UTF_8);
    }
}

File: Utils/src/main/kotlin/org/tobi29/scapes/engine/utils/io/ReadableByteStream.kt
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.utils.io

import org.tobi29.scapes.engine.utils.ByteBuffer
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

interface ReadableByteStream {
    fun available(): Int

    fun hasAvailable(): Boolean {
        return available() > 0
    }

    @Throws(IOException::class)
    fun skip(len: Int) {
        get(ByteBuffer(len))
    }

    @Throws(IOException::class)
    fun skip(len: Long) {
        while (len > Int.MAX_VALUE) {
            skip(Int.MAX_VALUE)
        }
        skip(len.toInt())
    }

    @Throws(IOException::class)
    operator fun get(buffer: ByteBuffer): ReadableByteStream {
        return get(buffer, buffer.remaining())
    }

    @Throws(IOException::class)
    operator fun get(buffer: ByteBuffer,
                     len: Int): ReadableByteStream

    @Throws(IOException::class)
    fun getSome(buffer: ByteBuffer): Boolean {
        return getSome(buffer, buffer.remaining())
    }

    @Throws(IOException::class)
    fun getSome(buffer: ByteBuffer,
                len: Int): Boolean

    @Throws(IOException::class)
    operator fun get(src: ByteArray,
                     off: Int = 0,
                     len: Int = src.size): ReadableByteStream {
        return get(ByteBuffer.wrap(src, off, len))
    }

    @Throws(IOException::class)
    fun getSome(src: ByteArray,
                off: Int,
                len: Int): Int {
        val buffer = ByteBuffer.wrap(src, off, len)
        val available = getSome(buffer)
        var position = buffer.position()
        position -= off
        if (position == 0 && !available) {
            return -1
        }
        assert(position <= len)
        return position
    }

    val boolean: Boolean
        @Throws(IOException::class)
        get() = get().toInt() != 0

    @Throws(IOException::class)
    fun get(): Byte

    val uByte: Short
        @Throws(IOException::class)
        get() {
            var value = get().toShort()
            if (value < 0) {
                value = (value + 0x100).toShort()
            }
            return value
        }

    val short: Short

    val uShort: Int
        @Throws(IOException::class)
        get() {
            var value = short.toInt()
            if (value < 0) {
                value += 0x10000
            }
            return value
        }

    val int: Int

    val uInt: Long
        @Throws(IOException::class)
        get() {
            var value = int.toLong()
            if (value < 0) {
                value += 0x100000000L
            }
            return value
        }

    val long: Long

    val float: Float

    val double: Double

    val byteArray: ByteArray
        @Throws(IOException::class)
        get() = getByteArray(Int.MAX_VALUE)

    @Throws(IOException::class)
    fun getByteArray(limit: Int): ByteArray {
        var len = uByte.toInt()
        if (len == 0xFF) {
            len = int
        }
        if (len < 0 || len > limit) {
            throw IOException(
                    "Array length outside of 0 to $limit: $limit")
        }
        val array = ByteArray(len)
        get(array, 0, array.size)
        return array
    }

    val byteArrayLong: ByteArray
        @Throws(IOException::class)
        get() = getByteArrayLong(Int.MAX_VALUE)

    @Throws(IOException::class)
    fun getByteArrayLong(limit: Int): ByteArray {
        var len = uShort
        if (len == 0xFFFF) {
            len = int
        }
        if (len < 0 || len > limit) {
            throw IOException(
                    "Array length outside of 0 to $limit: $limit")
        }
        val array = ByteArray(len)
        get(array, 0, array.size)
        return array
    }

    val string: String
        @Throws(IOException::class)
        get() = getString(Int.MAX_VALUE)

    @Throws(IOException::class)
    fun getString(limit: Int): String {
        return String(getByteArray(limit), StandardCharsets.UTF_8)
    }
}
